File: pkg/IP/ip.go
package IP

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
)

type IP struct {
	Value uint32 `json:"value"`
}

var ErrInvalidIPFormat = errors.New("invalid ip format")

func newIP(srcIP string) (IP, error) {
	if !IsValidIP(srcIP) {
		return IP{}, ErrInvalidIPFormat
	}

	parts := strings.Split(srcIP, ".")
	var ip uint32
	for _, part := range parts {
		p, err := strconv.Atoi(part)
		if err != nil {
			// this error should be impossible due to prior validation
			return IP{}, err
		}
		ip = ip<<8 | uint32(p)
	}

	return IP{Value: ip}, nil
}

func (ip IP) toString() string {
	// Extract each byte from the uint32 IP value
	byte1 := ip.Value >> 24 & 0xFF
	byte2 := ip.Value >> 16 & 0xFF
	byte3 := ip.Value >> 8 & 0xFF
	byte4 := ip.Value & 0xFF

	// Format the bytes into the standard IP address format
	return fmt.Sprintf("%d.%d.%d.%d", byte1, byte2, byte3, byte4)
}

File: pkg/IP/ip.py
from dataclasses import dataclass
from pkg.IP.validIP import is_valid_ip

class IPException(Exception):
    """Custom exception for invalid IP inputs."""
    pass

# Error for invalid IP format
ERR_INVALID_IP_FORMAT = IPException("invalid IP format")

@dataclass
class IP:
    value: int

    def to_string(self) -> str:
        """Convert the IP object to a standard string representation (e.g., '192.168.0.1')."""
        byte1 = (self.value >> 24) & 0xFF
        byte2 = (self.value >> 16) & 0xFF
        byte3 = (self.value >> 8) & 0xFF
        byte4 = self.value & 0xFF
        return f"{byte1}.{byte2}.{byte3}.{byte4}"

    @staticmethod
    def new_from_string(src_ip: str) -> 'IP':
        """
        Create a new IP object from the given string representation of an IP address.
        """
        if not is_valid_ip(src_ip):
            raise ERR_INVALID_IP_FORMAT

        parts = src_ip.split(".")
        ip_value = 0
        for part in parts:
            ip_value = (ip_value << 8) | int(part)

        return IP(value=ip_value)
